controllers/extension-certificate-service/pkg/controller/certservice/actuator.go

```diff
@@ -0,0 +1,335 @@
+
+	if err := a.deleteRBAC(ctx, kubecfg); err != nil {
+		return err
+	}
+
+	if err := a.client.Delete(ctx, secret); err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("error cleaning up secret for rbac manager: %v", err)
+	}
+
+	return nil
+}
+
+// InjectConfig injects the rest config to this actuator.
+func (a *actuator) InjectConfig(config *rest.Config) error {
+	a.config = config
+	return nil
+}
+
+// InjectClient injects the controller runtime client into the reconciler.
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+var certBrokerChartPath = filepath.Join(utils.ChartsPath, certBrokerName)
+
+const certBrokerName = "cert-broker"
+
+func (a *actuator) createCertBroker(ctx context.Context, shoot *gardenv1beta1.Shoot, namespace string) error {
+	shootDomain := shoot.Spec.DNS.Domain
+	if shootDomain == nil {
+		return fmt.Errorf("no domain given for shoot %s/%s", shoot.GetName(), shoot.GetNamespace())
+	}
+
+	var (
+		dns        []map[string]string
+		configSpec = a.certServiceConfig.Spec
+	)
+
+	for _, route53Provider := range configSpec.Providers.Route53 {
+		route53values := internal.CreateDNSProviderValue(&route53Provider, *shootDomain)
+		if route53values != nil {
+			dns = append(dns, route53values)
+		}
+	}
+	for _, cloudDNSProvider := range configSpec.Providers.CloudDNS {
+		cloudDNSValues := internal.CreateDNSProviderValue(&cloudDNSProvider, *shootDomain)
+		dns = append(dns, cloudDNSValues)
+	}
+
+	shootKubeconfig, err := a.createKubeconfigForCertManager(ctx, namespace)
+	if err != nil {
+		return err
+	}
+
+	shootKubeconfigChecksum := util.ComputeSecretCheckSum(shootKubeconfig.Data)
+
+	certBrokerConfig := map[string]interface{}{
+		"replicas": util.GetReplicaCount(shoot, 1),
+		"certbroker": map[string]interface{}{
+			"targetClusterSecret": shootKubeconfig.GetName(),
+		},
+		"certmanager": map[string]interface{}{
+			"clusterissuer": configSpec.IssuerName,
+			"dns":           dns,
+		},
+		"podAnnotations": map[string]interface{}{
+			"checksum/secret-cert-broker": shootKubeconfigChecksum,
+		},
+	}
+
+	certBrokerConfig, err = imagevector.ImageVector().InjectImages(certBrokerConfig, "", "", certBrokerName)
+	if err != nil {
+		return fmt.Errorf("failed to find image version for %s: %v", certBrokerName, err)
+	}
+
+	applier, err := kubernetes.NewChartApplierForConfig(a.config)
+	if err != nil {
+		return fmt.Errorf("failed to create chart applier: %v", err)
+	}
+
+	a.logger.Info("Component is being applied", "component", "cert-broker", "namespace", namespace)
+	return applier.ApplyChartInNamespace(
+		ctx,
+		certBrokerChartPath,
+		namespace,
+		certBrokerName,
+		certBrokerConfig,
+		nil,
+	)
+}
+
+func (a *actuator) deleteCertBroker(ctx context.Context, namespace string) error {
+	chartRenderer, err := chartrenderer.NewForConfig(a.config)
+	if err != nil {
+		return err
+	}
+	configValues := make(map[string]interface{})
+	release, err := chartRenderer.Render(
```

Comment:
Why not using [already existing utilities](https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-aws/pkg/controller/controlplane/actuator.go#L101-L152) instead of re-implementing some parts here?

Reply:
From my point of view the referenced utility isn't quite useful for the use case here.
It needs to know the `Objects    []*Object` you want to remove (see [Delete implementation](https://github.com/gardener/gardener-extensions/blob/master/vendor/github.com/gardener/gardener/pkg/utils/chart/chart.go#L117)), i.e. we could even use `client.Delete` directly.

OTOH, the coding below is very generic and can deal with future changes in the chart (e.g. manifest is added / removed).

Reply:
> OTOH, the coding below is very generic and can deal with future changes in the chart (e.g. manifest is added / removed).

I guess the coding below does not work in all cases. As an example, let's take a Helm chart that only deploys certain resources if a special flag in its `values.yaml` is set (we have a lot of these charts in the gardener repo itself). Here, you are again rendering the chart, however, with the default values (`configValues=nil`). That means that you could miss such manifests and they would remain orphaned in the system. The only way to prevent this is to have the EXACT same values for the deletion that you used during the time you deployed the chart, and these values you are not having.

> It needs to know the Objects []*Object you want to remove (see Delete implementation), i.e. we could even use client.Delete directly.

The reason why it needs to know the objects is what I just described above.
In this case you know exactly what is part of the `cert-broker` chart (maybe there are no such manifests that are only deployed if a certain field is set). However, the chances are high that if such a thing is introduced this deletion here does not work anymore.

I suggest to either use the utility to ease the cleaning/deletion process, or to simply call `client.Delete` as you suggested. The way it is currently implemented tries to solve it generically but this does not work in all cases. The next argument would be that you know the `cert-broker` chart and no such thing exists, but then it's again easier/cleaner to simply call `client.Delete`.